### ISSUE TITLE
Working With Forms Guide

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#3708]
__tldr;__ We propose a guide that will teach people how to work with forms, including:

- The different types of form elements, including custom elements.
- Using types and setters on ViewModels.
- Converters
- Validation
- Backup (or clone) and restore
- Custom events and attributes

> This was discussed on a [recent live stream (30:02)](https://youtu.be/ieZJrPwSZYQ?t=30m02s)

## The Problem

It's difficult to discover all the form-related tools that CanJS provides.  It is even more difficult to know how to combine them to achieve some desired functionality.

For example, the following are some useful form related tools:

- Custom Events
   - [can-event-dom-enter](https://canjs.com/doc/can-event-dom-enter.html) Listen to when someone hits enter in a form element
   - [can-event-dom-radiochange](https://canjs.com/doc/can-event-dom-radiochange.html) Listen to when a "group" of radio elements changes which is selected.
- Custom Attributes
   - [values](https://canjs.com/doc/can-util/dom/attr/attr.special.values.html) Used to work better with `<select multiple>`.  Allows reading and setting the values in the `<select>` as an array.
  - [focused](https://canjs.com/doc/can-util/dom/attr/attr.special.focused.html) - an attribute that indicates if the element is focused or not.
- [can-stache-bindings](https://canjs.com/doc/can-stache-bindings.html) basic binding syntaxes: `on:event`, `key:from`, `key:to`, and `key:bind`.
- can-stache-bindings "advanced" binding syntaxes: `on:input:value:to`
- `type: "number"` on DefineMap to ensure strings are converted to numbers
- [stringToAny](https://canjs.com/doc/can-util/js/string-to-any/string-to-any.html) utility that converts a string to what we judge is its best type.
- [can-stache-converters](https://canjs.com/doc/can-stache-converters.html)  that translate between two computes, including the [string-to-any](https://canjs.com/doc/can-stache-converters.string-to-any.html) converter.
- [can-define-validate-validatejs](https://canjs.com/doc/can-define-validate-validatejs.html) - validations on a DefineMap
- [can-connect-cloneable](https://canjs.com/doc/can-connect-cloneable.html) - Cloneable maps, whose clones can be changed, and the changes `.save()`d to the server and used to update the source map.

## The solution

I propose a guide that walks through the following use cases (or links to examples elsewhere in CanJS):

- Connecting a `viewModel` to all commonly implemented [form](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form) types (reading and writing).
  - Using converters and `DefineMap` types to keep your DOM as strings, but your VM as "nice" types.
  - Using custom attributes to simplify difficult to read form values.
- The different ways of connecting to those form types: 
  - Two-way binding pattern: `value:bind="key"`
  - Discrete event-up / value-down pattern: `on:input="updateKey(scope.element.value)" value:from="key"`
    - Q: What to do if user types, but you want to set `<input>` value "back" to old key value, but `key` hasn't changed? 
  - Hybrid and shortcut patterns: `on:input:value:to`
- Performing Validations
- Creating new data
- Updating data

**Comment by justinbmeyer** [comment]
> Q: What to do if user types, but you want to set <input> value "back" to old key value, but key hasn't changed?

This, imo, is really only solved well by streams which can produce the same value over and over again.  I think this problem is worth an issue and investigation of its own.

**Comment by cherifGsoul** [comment]
It would be great if the guide will have how to handle models relations, example ```Post``` map has many ```Category``` maps (Category.List) than how to handle ```categoryIDs``` <=>```Category.List```?
I'm assuming that ```Post``` and ```Category``` are models with ```can-connect```:
```js
var Category = DefineMap.extend({
  id: 'any',
  name :'string'
});

var Post = DefineMap.extend({
  id: 'any',
  title: 'string',
  categories: {
    Tye: Category.List,
    Value: Category.List
  }
});

var EditPostVM = DefineMap.extend({
  post: {
    Type: Post,
    Value: Post
  },
  categoryIds: {
    Type: DefineList
  },
  /**
   * for categories list
   */
  categoriesPromise: {
    value: function() {
      return Category.getList({})
    }
  },
  addRemoveCategoryId:function(ev) {
     //...
  },
  savePromise: function() {
    // ....
  }
});

//stache
...
{{#if categoriesPromise.isResolved}}
 {{#each categoriesPromise.value}}
    <label>
      <input type="checkbox" value:from="id"  on:change="addRemoveCategoryId(scope.event)">
    </label>
 {{/each}}
{{/if}}
....
```

**Comment by frank-dspeed** [comment]
It would be more then great if we show handling of file uploads i mean the special file propertys. how they get binded best.

**Comment by phillipskevin** [comment]
Here is the outline for the first draft of the guide that I am working on now:

	## Form Elements
	
	### <input type="text">
	* binding to events
	  - on:change
	  - list of standard events that can be bound to
	* binding to attributes
	  - value:to
	  - value:from
	  - value=“{{foo}}”
	  - list of standard attributes that can be bound to
	* binding to events and attributes
	  - value:bind
	  - on:input:value:to
	
	
	### <input type="checkbox">
	* checked:from
	* {{#if checked}}checked{{/if}}
	* checked:bind=“boolean-to-inList(key, list)”
	* checked:bind=“either-or(…)” 
	
	
	### <input type="radio">
	* can-event-dom-radiochange
	* checked:bind=“equal(key, value)” 
	
	
	### <input type="number">
	* type: "number"
	
	
	### <input type="submit">
	* can-event-dom-enter
	
	
	### <input type="file">
	
	
	### <select>
	
	
	### <select multiple>
	* value:bind
	
	
	### <textarea>
	
	
	## Fancy bindings
	
	### Non-standard attributes
	* focused
	* disabled
	
	
	### Non-standard events
	* Reasoning
	  - Keep DOM-related functionality out of your ViewModel (ev.type === 13, etc)
	* enter
	  - can-event-dom-enter
	* radiochange
	  - can-event-dom-radiochange
	
	
	### events up, data down
	* on:change=“scope.set(“key”, scope.element.value)” value:from=“key”
	
	
	### Resetting <input> value back
	* explain that two-way binding normally sets <input> value back when a “change” event occurs on the ViewModel property
	* if a setter or resolver does not change the value (because it is illegal), there is no change event
	* with :bind, can-stache-bindings will handle this for you
	
	## Validation
	* can-define-validate-validatejs

**Comment by phillipskevin** [comment]
I'm leaving out the `can-event-dom-radiochange` for now as I don't see a use-case for using this directly (not through can-stache-bindings and the `checked` attribute).

**Comment by frank-dspeed** [comment]
@phillipskevin can-connect-cloneable isn't that replacing can-define-map-backup ? stuff?

**Comment by phillipskevin** [comment]
I think `can-define-backup` and `can-connect-cloneable` have slightly different goals, but they are related. I don't think one is replacing the other.

**Comment by phillipskevin** [comment]
Remaining changes:
- [X] Remove use of `<label>` where they prevent you from clicking outside `<input>` to produce `"change"` events
- [ ] Revamp overview section so that it gives people a better understanding of what they're getting into (take a look at Tech Overview guide)
- [X] Remove use of `scope.set(...)` in "data down, actions up" section
- [X] Expand "data down, actions up" section to show an example with multiple components
- [X] Use Components in all the examples
- [ ] Add a "working with related data" example
  - make, model, year example with can-ajax
  - how to load data
  - how to use promises in forms
  - the `string-to-any` utility
  - link to issue on "creating new data" / "updating data"
- [X] explain when `on:click` will be called when hitting the Enter key inside an `<input>` in a form
- [ ] Make Custom Attributes and Custom Events more obvious
